### PR TITLE
rdcore/kargs: call `zipl` on s390x after editing BLS

### PR DESCRIPTION
**Pull request by jlebon**
Otherwise the changed kargs won't take effect.

### CODE REVIEW

File: src/bin/rdcore/kargs.rs
Comment (jlebon) [inline src/bin/rdcore/kargs.rs:37]: Not 100% sure if that's all we need. /cc @nikita-dubrovskii

Basing this off of `ignition-firstboot-complete`, which just calls `zipl`, except that here the boot partition isn't necessarily mounted on `/boot` so we need to point `zipl` at it: https://github.com/coreos/ignition/blob/3885b9f07c96baf7dcc12c85b914ded151c8fd6e/systemd/ignition-firstboot-complete.service#L26

Comment (bgilbert) [reply src/bin/rdcore/kargs.rs:37]: Also, if a bare `zipl` works, I'm confused by all the explicit arguments in `install_bootloader()`.  If zipl indeed knows how to read BLS configs, could it do that in the `install_bootloader()` case too?

Comment (nikita-dubrovskii) [reply src/bin/rdcore/kargs.rs:37]: I'll check it.

Comment (nikita-dubrovskii) [reply src/bin/rdcore/kargs.rs:37]: So, `--target` option doesn't work without specifying `--image` and/or others. what could be used is `--blsdir` , but in such case `/boot` shoud be `rw` accessible.

Comment (jlebon) [reply src/bin/rdcore/kargs.rs:37]: > what could be used is `--blsdir` , but in such case `/boot` shoud be `rw` accessible. 

Huh interesting, does it actually need write access to `/boot`, or just for the boot partition to be mounted at `/boot`? If the former, what does it write there?

**Comment by jlebon** [comment]
OK, updated this now! One thing that became obvious is that we need a little more sophistication here because I want `rdcore kargs` to work equally well whether it's the first boot or not. This meant having the s390x code auto-detect it and doing the right thing.

@nikita-dubrovskii Would you be able to test this on s390x? Ideally using both `--firstboot-kargs` and an Ignition config which uses Ignition kargs.

**Comment by nikita-dubrovskii** [comment]
Thanks @jschintag!

**Comment by jlebon** [comment]
OK weird, I really thought `zipl` was already in the initramfs for FCOS at this point. @nikita-dubrovskii Can you submit that patch to the f-c-c repo?

**Comment by nikita-dubrovskii** [comment]
> OK weird, I really thought `zipl` was already in the initramfs for FCOS at this point. @nikita-dubrovskii Can you submit that patch to the f-c-c repo?

Done - https://github.com/coreos/fedora-coreos-config/pull/1146

**Comment by nikita-dubrovskii** [comment]
LGTM

**Comment by jlebon** [comment]
Updated now to drop the tempdir move! (Good catch!) And also added @nikita-dubrovskii's `zipl` dracut patch here.

Because I'm very wary of code I can't test (I did compile-test it at least by fudging with the conditionals), I would appreciate still a final test of this PR using `--firstboot-kargs` and also using Ignition kargs.

**Comment by nikita-dubrovskii** [comment]
> I would appreciate still a final test of this PR using `--firstboot-kargs` and also using Ignition kargs.

done, works well

**Comment by jlebon** [comment]
Awesome, thanks! Will let @bgilbert have a final look before we merge.

File: src/s390x/zipl.rs
Comment (bgilbert) [inline src/s390x/zipl.rs:34]: Stale comment.

File: src/s390x/zipl.rs
Comment (bgilbert) [inline src/s390x/zipl.rs:84]: The `ignition.firstboot` file [isn't that simple](https://github.com/coreos/coreos-installer/blob/d14fe5dbb5a864eee4d45070a6a7eb3943b54249/src/install.rs#L307), unfortunately.

File: src/s390x/mod.rs
Comment (bgilbert) [inline src/s390x/mod.rs:20]: The distinction between these two functions isn't immediately obvious from the names.  Can we rename one or both?

**Review by bgilbert — COMMENTED** [review]
LGTM generally, and this is a nice cleanup, but I think `--firstboot-kargs` will put some extra garbage in the kernel command line.

File: src/s390x/zipl.rs
Comment (jlebon) [reply src/s390x/zipl.rs]: Ahhh of course. There's no easy way to solve this generically, but at least we can special-case the format we expect. Fix added!

File: src/s390x/mod.rs
Comment (jlebon) [reply src/s390x/mod.rs]: Added another commit that cleans things up a bit more.

File: src/s390x/zipl.rs
Comment (bgilbert) [inline src/s390x/zipl.rs:134]: Other things to possibly test:
- Valid `set` line followed by a newline
- Valid `set` line preceded by something else
- Valid `set` line followed by something else

**Review by bgilbert — APPROVED** [review]

**Review by bgilbert — APPROVED** [review]

**Comment by jschintag** [comment]
Thanks for the work. It works on s390x.

**Pull request merged**
